mesonbuild/mparser.py

```diff
@@ -240,24 +240,25 @@ def lex(self, filename: str) -> T.Generator[Token, None, None]:
             if not matched:
                 raise ParseException('lexer', self.getline(line_start), lineno, col)
 
-@dataclass(eq=False)
+@dataclass
 class BaseNode:
     lineno: int
     colno: int
-    filename: str
-    end_lineno: T.Optional[int] = None
-    end_colno: T.Optional[int] = None
+    filename: str = field(hash=False)
+    end_lineno: int = field(hash=False)
+    end_colno: int = field(hash=False)
 
-    def __post_init__(self) -> None:
-        if self.end_lineno is None:
-            self.end_lineno = self.lineno
-        if self.end_colno is None:
-            self.end_colno = self.colno
+    def __init__(self, lineno: int, colno: int, filename: str, end_lineno: T.Optional[int] = None, end_colno: T.Optional[int] = None) -> None:
+        self.lineno = lineno
+        self.colno = colno
+        self.filename = filename
+        self.end_lineno = end_lineno if end_lineno is not None else lineno
+        self.end_colno = end_colno if end_colno is not None else colno
```

Comment:
I think you can keep using the `__post_init__` here by doing something like:

```python
end_lineno: InitVar[T.Optional[int]] = field(default=None, hash=False)
end_colno: InitVar[T.Optional[int]] = field(default=None, hash=False)

def __post_init__(self, end_lineno: T.Optional[int], end_colno: T.Optional[int]):
    self.end_lineno = end_lineno if end_lineno is not None else self.lineno
    self.end_colno = end_colno if end_colno is not None else self.colno
```

But it's up to you which you prefer I don't have a strong opinion

Reply:
mypy does not seems to like that solution. All other node classes have `__init__` too, so I don't think it's a big issue to keep it here.

Was that your only concern on this PR?

Reply:
oh, I see you approved the PR already, LGTM.